TypeScript error in FakeRTCDataChannel

```
src/FakeRTCDataChannel.ts:23:14 - error TS2420: Class 'FakeRTCDataChannel' incorrectly implements interface 'RTCDataChannel'.
  Types of property 'addEventListener' are incompatible.
    Type '{ <T extends string>(type: T, callback?: EventListener<this, Event<string>> | null | undefined, options?: AddOptions | undefined): void; (type: string, callback?: EventListener<...> | ... 1 more ... | undefined, options?: AddOptions | undefined): void; <T extends string>(type: T, callback: EventListener<...> | ... 1...' is not assignable to type '{ <K extends keyof RTCDataChannelEventMap>(type: K, listener: (this: RTCDataChannel, ev: RTCDataChannelEventMap[K]) => any, options?: boolean | ... 1 more ... | undefined): void; (type: string, listener: EventListenerOrEventListenerObject, options?: boolean | ... 1 more ... | undefined): void; }'.
      Types of parameters 'callback' and 'listener' are incompatible.
        Type '(this: RTCDataChannel, ev: any) => any' is not assignable to type 'EventListener<this, Event<string>> | null | undefined'.
          Type '(this: RTCDataChannel, ev: any) => any' is not assignable to type 'CallbackFunction<this, Event<string>>'.
            The 'this' types of each signature are incompatible.
              Type 'this' is not assignable to type 'RTCDataChannel'.
                Type 'FakeRTCDataChannel' is not assignable to type 'RTCDataChannel'.
                  Types of property 'addEventListener' are incompatible.
                    Type '{ <T extends string>(type: T, callback?: EventListener<this, Event<string>> | null | undefined, options?: AddOptions | undefined): void; (type: string, callback?: EventListener<...> | ... 1 more ... | undefined, options?: AddOptions | undefined): void; <T extends string>(type: T, callback: EventListener<...> | ... 1...' is not assignable to type '{ <K extends keyof RTCDataChannelEventMap>(type: K, listener: (this: RTCDataChannel, ev: RTCDataChannelEventMap[K]) => any, options?: boolean | ... 1 more ... | undefined): void; (type: string, listener: EventListenerOrEventListenerObject, options?: boolean | ... 1 more ... | undefined): void; }'.
                      Types of parameters 'callback' and 'listener' are incompatible.
                        Type '(this: RTCDataChannel, ev: any) => any' is not assignable to type 'EventListener<this, Event<string>> | null | undefined'.

23 export class FakeRTCDataChannel extends EventTarget implements RTCDataChannel
```

This is currently "workarounded" with a `@ts-ignore` above `FakeRTCDataChannel` class definition and also another 2 `@ts-ignore` in `Handler.ts`, which is hell ofc.

Problem is that **mediasoup-client** relies on "dom" TS library, and here we are defining `FakeRTCDataChannel` which uses **event-target-shim** dep to mimics DOM's `EventTarget`. Related task in **mediasoup-client**: https://github.com/versatica/mediasoup-client/issues/279

--- a/package.json
+++ b/package.json
@@ -54,24 +54,23 @@
 		"release": "node npm-scripts.mjs release"
 	},
 	"dependencies": {
-		"@types/debug": "^4.1.12",
 		"debug": "^4.4.0",
-		"event-target-shim": "^6.0.2",
 		"fake-mediastreamtrack": "^1.2.0",
-		"mediasoup-client": "^3.9.4",
+		"mediasoup-client": "^3.9.5",
 		"netstring": "^0.3.0",
 		"sdp-transform": "^2.15.0",
 		"uuid": "^11.1.0"
 	},
 	"devDependencies": {
-		"@eslint/js": "^9.22.0",
+		"@eslint/js": "^9.23.0",
 		"@types/jest": "^29.5.14",
+		"@types/debug": "^4.1.12",
 		"@types/node": "^22.13.11",
 		"@types/sdp-transform": "^2.4.9",
 		"@types/uuid": "^10.0.0",
 		"@typescript-eslint/eslint-plugin": "^8.27.0",
 		"@typescript-eslint/parser": "^8.27.0",
-		"eslint": "^9.22.0",
+		"eslint": "^9.23.0",
 		"eslint-config-prettier": "^10.1.1",
 		"eslint-plugin-jest": "^28.11.0",
 		"eslint-plugin-prettier": "^5.2.3",

--- a/src/AiortcMediaStream.ts
+++ b/src/AiortcMediaStream.ts
@@ -1,20 +1,12 @@
 import { v4 as uuidv4 } from 'uuid';
-import {
-	EventTarget,
-	Event,
-	getEventAttributeValue,
-	setEventAttributeValue,
-} from 'event-target-shim';
 import { FakeMediaStreamTrack } from 'fake-mediastreamtrack';
 
 export class AiortcMediaStream extends EventTarget implements MediaStream {
 	readonly #id: string;
 	readonly #tracks: Map<string, FakeMediaStreamTrack> = new Map();
-
-	// Event listeners. These are cosmetic public members to make TS happy.
-	// NOTE: We never emit these events.
-	// public onaddtrack: (this: AiortcMediaStream, ev: Event) => any;
-	// public onremovetrack: (this: MediaStream, ev: Event) => any;
+	// Events.
+	#onaddtrack: ((this: AiortcMediaStream, ev: Event) => any) | null = null;
+	#onremovetrack: ((this: AiortcMediaStream, ev: Event) => any) | null = null;
 
 	constructor(tracks: FakeMediaStreamTrack[]) {
 		super();
@@ -36,20 +28,40 @@ export class AiortcMediaStream extends EventTarget implements MediaStream {
 		);
 	}
 
-	get onaddtrack(): any {
-		return getEventAttributeValue(this, 'addtrack');
+	get onaddtrack(): ((this: MediaStream, ev: Event) => any) | null {
+		return this.#onaddtrack as ((this: MediaStream, ev: Event) => any) | null;
 	}
 
-	set onaddtrack(listener) {
-		setEventAttributeValue(this, 'addtrack', listener);
+	set onaddtrack(handler: ((this: MediaStream, ev: Event) => any) | null) {
+		if (this.#onaddtrack) {
+			this.removeEventListener('addtrack', this.#onaddtrack);
+		}
+
+		this.#onaddtrack = handler;
+
+		if (handler) {
+			this.addEventListener('addtrack', handler);
+		}
 	}
 
-	get onremovetrack(): any {
-		return getEventAttributeValue(this, 'removetrack');
+	get onremovetrack():
+		| ((this: MediaStream, ev: MediaStreamTrackEvent) => any)
+		| null {
+		return this.#onremovetrack as
+			| ((this: MediaStream, ev: Event) => any)
+			| null;
 	}
 
-	set onremovetrack(listener) {
-		setEventAttributeValue(this, 'removetrack', listener);
+	set onremovetrack(handler: ((this: MediaStream, ev: Event) => any) | null) {
+		if (this.#onremovetrack) {
+			this.removeEventListener('removetrack', this.#onremovetrack);
+		}
+
+		this.#onremovetrack = handler;
+
+		if (handler) {
+			this.addEventListener('removetrack', handler);
+		}
 	}
 
 	/**
@@ -79,10 +91,8 @@ export class AiortcMediaStream extends EventTarget implements MediaStream {
 		return Array.from(this.#tracks.values());
 	}
 
-	// @ts-expect-error --- // TypeScript things that mediaStream.getTrackById()
-	// should return null instead of undefined. It's wrong.
-	getTrackById(trackId: string): FakeMediaStreamTrack | undefined {
-		return this.#tracks.get(trackId);
+	getTrackById(trackId: string): FakeMediaStreamTrack | null {
+		return this.#tracks.get(trackId) ?? null;
 	}
 
 	// eslint-disable-next-line @typescript-eslint/no-unused-vars

--- a/src/FakeRTCDataChannel.ts
+++ b/src/FakeRTCDataChannel.ts
@@ -1,9 +1,3 @@
-import {
-	EventTarget,
-	Event,
-	getEventAttributeValue,
-	setEventAttributeValue,
-} from 'event-target-shim';
 import { Logger } from './Logger';
 import { Channel } from './Channel';
 import { InvalidStateError } from './errors';
@@ -19,7 +13,6 @@ export type FakeRTCDataChannelOptions = {
 	protocol?: string;
 };
 
-// @ts-expect-error --- TODO:
 // https://github.com/versatica/mediasoup-client-aiortc/issues/24
 export class FakeRTCDataChannel extends EventTarget implements RTCDataChannel {
 	// Internal data.
@@ -27,19 +20,25 @@ export class FakeRTCDataChannel extends EventTarget implements RTCDataChannel {
 	// Channel.
 	readonly #channel: Channel;
 	// Members for RTCDataChannel standard public getters/setters.
-	#id: number;
-	#negotiated = true; // mediasoup just uses negotiated DataChannels.
-	#ordered: boolean;
-	#maxPacketLifeTime: number | null;
-	#maxRetransmits: number | null;
-	#label: string;
-	#protocol: string;
+	readonly #id: number;
+	readonly #negotiated = true; // mediasoup just uses negotiated DataChannels.
+	readonly #ordered: boolean;
+	readonly #maxPacketLifeTime: number | null;
+	readonly #maxRetransmits: number | null;
+	readonly #label: string;
+	readonly #protocol: string;
 	#readyState: RTCDataChannelState = 'connecting';
-	#bufferedAmount = 0;
-	#bufferedAmountLowThreshold = 0;
+	#bufferedAmount;
+	#bufferedAmountLowThreshold;
 	#binaryType: BinaryType = 'arraybuffer';
-	// NOTE: Deprecated as per spec, but still required by TS/ RTCDataChannel.
-	#priority: RTCPriorityType = 'high';
+	// Events.
+	#onopen: ((this: FakeRTCDataChannel, ev: Event) => any) | null = null;
+	#onclosing: ((this: FakeRTCDataChannel, ev: Event) => any) | null = null;
+	#onclose: ((this: FakeRTCDataChannel, ev: Event) => any) | null = null;
+	#onmessage: ((this: FakeRTCDataChannel, ev: Event) => any) | null = null;
+	#onbufferedamountlow: ((this: FakeRTCDataChannel, ev: Event) => any) | null =
+		null;
+	#onerror: ((this: FakeRTCDataChannel, ev: Event) => any) | null = null;
 
 	constructor(
 		internal: { handlerId: string; dataChannelId: string },
@@ -138,61 +137,108 @@ export class FakeRTCDataChannel extends EventTarget implements RTCDataChannel {
 		logger.warn('binaryType setter not implemented, using "arraybuffer"');
 	}
 
-	// NOTE: Deprecated in the spec but required by RTCDataChannel TS definition.
-	get priority(): RTCPriorityType {
-		return this.#priority;
+	get onopen(): ((this: RTCDataChannel, ev: Event) => any) | null {
+		return this.#onopen as ((this: RTCDataChannel, ev: Event) => any) | null;
 	}
 
-	set priority(value: RTCPriorityType) {
-		this.#priority = value;
+	set onopen(handler: ((this: FakeRTCDataChannel, ev: Event) => any) | null) {
+		if (this.#onopen) {
+			this.removeEventListener('open', this.#onopen);
+		}
+
+		this.#onopen = handler;
+
+		if (handler) {
+			this.addEventListener('open', handler);
+		}
 	}
 
-	get onopen(): any {
-		return getEventAttributeValue(this, 'open');
+	get onclosing(): ((this: RTCDataChannel, ev: Event) => any) | null {
+		return this.#onclosing as ((this: RTCDataChannel, ev: Event) => any) | null;
 	}
 
-	set onopen(listener) {
-		setEventAttributeValue(this, 'open', listener);
+	set onclosing(
+		handler: ((this: FakeRTCDataChannel, ev: Event) => any) | null
+	) {
+		if (this.#onclosing) {
+			this.removeEventListener('closing', this.#onclosing);
+		}
+
+		this.#onclosing = handler;
+
+		if (handler) {
+			this.addEventListener('closing', handler);
+		}
 	}
 
-	get onclosing(): any {
-		return getEventAttributeValue(this, 'closing');
+	get onclose(): ((this: RTCDataChannel, ev: Event) => any) | null {
+		return this.#onclose as ((this: RTCDataChannel, ev: Event) => any) | null;
 	}
 
-	set onclosing(listener) {
-		setEventAttributeValue(this, 'closing', listener);
+	set onclose(handler: ((this: FakeRTCDataChannel, ev: Event) => any) | null) {
+		if (this.#onclose) {
+			this.removeEventListener('close', this.#onclose);
+		}
+
+		this.#onclose = handler;
+
+		if (handler) {
+			this.addEventListener('close', handler);
+		}
 	}
 
-	get onclose(): any {
-		return getEventAttributeValue(this, 'close');
+	get onmessage(): ((this: RTCDataChannel, ev: Event) => any) | null {
+		return this.#onmessage as ((this: RTCDataChannel, ev: Event) => any) | null;
 	}
 
-	set onclose(listener) {
-		setEventAttributeValue(this, 'close', listener);
+	set onmessage(
+		handler: ((this: FakeRTCDataChannel, ev: Event) => any) | null
+	) {
+		if (this.#onmessage) {
+			this.removeEventListener('message', this.#onmessage);
+		}
+
+		this.#onmessage = handler;
+
+		if (handler) {
+			this.addEventListener('message', handler);
+		}
 	}
 
-	get onmessage(): any {
-		return getEventAttributeValue(this, 'message');
+	get onbufferedamountlow(): ((this: RTCDataChannel, ev: Event) => any) | null {
+		return this.#onbufferedamountlow as
+			| ((this: RTCDataChannel, ev: Event) => any)
+			| null;
 	}
 
-	set onmessage(listener) {
-		setEventAttributeValue(this, 'message', listener);
+	set onbufferedamountlow(
+		handler: ((this: FakeRTCDataChannel, ev: Event) => any) | null
+	) {
+		if (this.#onbufferedamountlow) {
+			this.removeEventListener('bufferedamountlow', this.#onbufferedamountlow);
+		}
+
+		this.#onbufferedamountlow = handler;
+
+		if (handler) {
+			this.addEventListener('bufferedamountlow', handler);
+		}
 	}
 
-	get onbufferedamountlow(): any {
-		return getEventAttributeValue(this, 'bufferedamountlow');
+	get onerror(): ((this: RTCDataChannel, ev: Event) => any) | null {
+		return this.#onerror as ((this: RTCDataChannel, ev: Event) => any) | null;
 	}
 
-	set onbufferedamountlow(listener) {
-		setEventAttributeValue(this, 'bufferedamountlow', listener);
-	}
+	set onerror(handler: ((this: FakeRTCDataChannel, ev: Event) => any) | null) {
+		if (this.#onerror) {
+			this.removeEventListener('error', this.#onerror);
+		}
 
-	get onerror(): any {
-		return getEventAttributeValue(this, 'error');
-	}
+		this.#onerror = handler;
 
-	set onerror(listener) {
-		setEventAttributeValue(this, 'error', listener);
+		if (handler) {
+			this.addEventListener('error', handler);
+		}
 	}
 
 	close(): void {
@@ -269,7 +315,6 @@ export class FakeRTCDataChannel extends EventTarget implements RTCDataChannel {
 					}
 
 					case 'message': {
-						// @ts-expect-error --- On purpose.
 						this.dispatchEvent(new MessageEvent('message', { data }));
 
 						break;
@@ -285,7 +330,6 @@ export class FakeRTCDataChannel extends EventTarget implements RTCDataChannel {
 						}
 
 						this.dispatchEvent(
-							// @ts-expect-error --- On purpose.
 							new MessageEvent('message', { data: arrayBuffer })
 						);
 

--- a/src/Handler.ts
+++ b/src/Handler.ts
@@ -753,8 +753,6 @@ export class Handler extends HandlerInterface {
 		};
 
 		return {
-			// @ts-expect-error --- TODO:
-			// https://github.com/versatica/mediasoup-client-aiortc/issues/24
 			dataChannel,
 			sctpStreamParameters,
 		};
@@ -1150,8 +1148,6 @@ export class Handler extends HandlerInterface {
 			this.#hasDataChannelMediaSection = true;
 		}
 
-		// @ts-expect-error --- TODO:
-		// https://github.com/versatica/mediasoup-client-aiortc/issues/24
 		return { dataChannel };
 	}
 

